File: my_distribution.py
import math
import sys
import dnest4


class MyDistribution:
    def __init__(self):
        self.center = 0.0
        self.width = 0.0
        self.mu = 0.0

    def from_prior(self, r):
        # Cauchy prior centered on 5.901 = log(365 days).
        self.center = 5.901 + math.tan(math.pi * (0.97 * dnest4.rand() - 0.485))
        self.width = 0.1 + 2.9 * dnest4.rand()
        self.mu = math.exp(math.tan(math.pi * (0.97 * dnest4.rand() - 0.485)))

    def perturb_hyperparameters(self, rng):
        logH = 0.0

        which = int(dnest4.rand(3))

        if which == 0:
            self.center = (math.atan(self.center - 5.901) / math.pi + 0.485) / 0.97
            self.center += dnest4.randh()
            self.center = 5.901 + math.tan(math.pi * (0.97 * self.center - 0.485))
        elif which == 1:
            self.width += 2.9 * dnest4.randh()
            self.width = self._wrap(self.width, 0.1, 3.0)
        else:
            self.mu = math.log(self.mu)
            self.mu = (math.atan(self.mu) / math.pi + 0.485) / 0.97
            self.mu += dnest4.randh()
            self.mu = math.tan(math.pi * (0.97 * self.mu - 0.485))
            self.mu = math.exp(self.mu)

        return logH

    def log_pdf(self, vec):
        if (
            vec[1] < 0.0
            or vec[2] < 0.0
            or vec[2] > 2.0 * math.pi
            or vec[3] < 0.0
            or vec[3] > 0.8189776
            or vec[4] < 0.0
            or vec[4] > 2.0 * math.pi
        ):
            return -1e300

        return (
            -math.log(2.0 * self.width)
            - abs(vec[0] - self.center) / self.width
            - math.log(self.mu)
            - vec[1] / self.mu
            + 2.1 * math.log(1.0 - vec[3] / 0.995)
        )

    def from_uniform(self, vec):
        if vec[0] < 0.5:
            self.center = self.center + self.width * math.log(2.0 * vec[0])
        else:
            self.center = self.center - self.width * math.log(2.0 - 2.0 * vec[0])
        vec[1] = -self.mu * math.log(1.0 - vec[1])
        vec[2] = 2.0 * math.pi * vec[2]
        vec[3] = 1.0 - math.pow(1.0 - 0.995 * vec[3], 1.0 / 3.1)
        vec[4] = 2.0 * math.pi * vec[4]

    def to_uniform(self, vec):
        if vec[0] < self.center:
            vec[0] = 0.5 * math.exp((vec[0] - self.center) / self.width)
        else:
            vec[0] = 1.0 - 0.5 * math.exp((self.center - vec[0]) / self.width)
        vec[1] = 1.0 - math.exp(-vec[1] / self.mu)
        vec[2] = vec[2] / (2.0 * math.pi)
        vec[3] = 1.0 - math.pow(1.0 - vec[3] / 0.995, 3.1)
        vec[4] = vec[4] / (2.0 * math.pi)

    def print(self, out=sys.stdout):
        print(self.center, self.width, self.mu, file=out)

    def _wrap(self, x, xmin, xmax):
        width = xmax - xmin
        while x < xmin:
            x += width
        while x >= xmax:
            x -= width
        return x


if __name__ == "__main__":
    import random

    # dnest4 = random.Random()
    distribution = MyDistribution()

    distribution.from_prior(None)
    print("Parameters after initialization:")
    distribution.print()

    vec = [0.3, 0.5, 1.0, 0.2, 1.5]
    print(f"Log PDF for vector {vec}: {distribution.log_pdf(vec)}")

File: MyDistribution.cpp
#include "MyDistribution.h"
#include <cmath>

MyDistribution::MyDistribution()
{

}

void MyDistribution::from_prior(DNest4::RNG& rng)
{
	// Cauchy prior centered on 5.901 = log(365 days).
	// 0.97 and 0.485 account for truncation
	// tan()... denotes cauchy dist, check the link:
	// https://en.wikipedia.org/wiki/Cauchy_distribution
	// these all are Hyperparams defined in the paper

	center = 5.901 + tan(M_PI*(0.97*rng.rand() - 0.485));  // median orbital period
	width = 0.1 + 12.9*rng.rand();  // diversity of orbital periods
	mu = exp(tan(M_PI*(0.97*rng.rand() - 0.485)));  // mean amplitude (m/s), exp of Cauchy
}

double MyDistribution::perturb_hyperparameters(DNest4::RNG& rng)
{
	double logH = 0.;

	int which = rng.rand_int(3);

	if(which == 0)
	{
		// this just updates center with rand gauss noise
		center = (atan(center - 5.901)/M_PI + 0.485)/0.97;
		center += rng.randh();
		DNest4::wrap(center, 0., 1.);  // cuts the param (which is basically rng.rand()) if too big or small
		center = 5.901 + tan(M_PI*(0.97*center - 0.485));
	}
	else if(which == 1)
	{
		// this updates width with rand gauss noise
		width += 12.9*rng.randh();
		DNest4::wrap(width, 0.1, 13.);
	}
	else
	{
		// this updates semi-amplitude with rand gauss noise
		mu = log(mu);
		mu = (atan(mu)/M_PI + 0.485)/0.97;
		mu += rng.randh();
		DNest4::wrap(mu, 0., 1.);
		mu = tan(M_PI*(0.97*mu - 0.485));
		mu = exp(mu);
	}
	return logH;
}

// vec[0] = "position" (log-period)
// vec[1] = amplitude
// vec[2] = phase
// vec[3] = v0 - this is eccentrincity
// vec[4] = viewing angle
// these all are the planet params defined in the paper

double MyDistribution::log_pdf(const std::vector<double>& vec) const
{
	if(vec[1] < 0. ||
			vec[2] < 0. || vec[2] > 2.*M_PI ||
			vec[3] < 0. || vec[3] > 0.8189776 ||
			vec[4] < 0. || vec[4] > 2.*M_PI)
		return -1E300;

	return  -log(2.*width) - abs(vec[0] - center)/width  // log of biexponential for Pi (paper, footnote, page 4)
		-log(mu) - vec[1]/mu  // log of exponential for Ai (with param 1/mu for some reason)
		+ 2.1*log(1. - vec[3]/0.995);  // log of beta for eccentricity
	// other distributions are not present since they are constant???
}

void MyDistribution::from_uniform(std::vector<double>& vec) const
// this transforms some uniformly distributed parameter vectors
// to the desired distributions described in table 1
{
	if(vec[0] < 0.5)
		vec[0] = center + width*log(2.*vec[0]);
	else
		vec[0] = center - width*log(2. - 2.*vec[0]);
	vec[1] = -mu*log(1. - vec[1]);
	vec[2] = 2.*M_PI*vec[2];
	vec[3] = 1. - pow(1. - 0.995*vec[3], 1./3.1);
	vec[4] = 2.*M_PI*vec[4];
}

void MyDistribution::to_uniform(std::vector<double>& vec) const
{
	if(vec[0] < center)
		vec[0] = 0.5*exp((vec[0] - center)/width);
	else
		vec[0] = 1. - 0.5*exp((center - vec[0])/width);
	vec[1] = 1. - exp(-vec[1]/mu);
	vec[2] = vec[2]/(2.*M_PI);
	vec[3] = 1. - pow(1. - vec[3]/0.995, 3.1);
	vec[4] = vec[4]/(2.*M_PI);
}

void MyDistribution::print(std::ostream& out) const
{
	out<<center<<' '<<width<<' '<<mu<<' ';
}

